File: Func/PaginatorScope.cs
using Func.Types;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Func
{
    public static class PaginatorScope
    {
        public static PaginatorState
            Init(
                IntMore0Less65535Exclsv pCurrentPage,
                IntMore0Less65535Exclsv pItemsPerPage,
                IntMore0Less65535Exclsv pPagesToSkip,
                int pTotalNumberOfItemsInDB
                )
        {
            var res = new PaginatorState();
            res.CurrentPage = pCurrentPage;
            res.ItemsPerPage = pItemsPerPage;
            res.PagesToSkip = pPagesToSkip;
            res.TotalNumberOfItemsInDB = pTotalNumberOfItemsInDB;

            return NextState(res);
        }

        private static PaginatorState
            NextState(PaginatorState pPaginatorState)
        {
            var res = pPaginatorState.
                      PipeForward(GetNumberOfPages()).
                      PipeForward(IsValidLeft()).
                      PipeForward(IsValidLeftMore()).
                      PipeForward(IsValidRight()).
                      PipeForward(IsValidRightMore());

            return res;
        }

        public static PaginatorState
            GoRight(
            this PaginatorState pState)
        {
            return NextState(pState.With(z => z.CurrentPage.Value++));
        }

        public static PaginatorState
            GoLeft(
            this PaginatorState pState)
        {
            return NextState(pState.With(z => z.CurrentPage.Value--));
        }

        public static PaginatorState
            GoRightMore(
            this PaginatorState pState)
        {
            return NextState(pState.With(z => z.CurrentPage.Value += pState.PagesToSkip.Value));
        }

        public static PaginatorState
            GoLeftMore(
            this PaginatorState pState)
        {
            return NextState(pState.With(z => z.CurrentPage.Value -= pState.PagesToSkip.Value));
        }

        private static Func<PaginatorState, uint>
            GetLeftIndex() => (pState) =>
                {
                    return (uint)((pState.CurrentPage.Value - 1) * pState.ItemsPerPage.Value);
                };

        private static Func<PaginatorState, uint>
            RightIndex => (pState) =>
                {
                    return (uint)(pState.CurrentPage.Value * pState.ItemsPerPage.Value);
                };

        public static Func<uint, uint, IEnumerable<int>, IEnumerable<int>>
            GetDataStartEndIndex() => (startIndex, endIndex, DbData) =>
                {
                    return DbData.Where(z => z > startIndex && z <= endIndex);
                };

        private static Func<PaginatorState, PaginatorState>
            GetNumberOfPages() => (pState) =>
                {
                    var res = (ushort)Math.Round((double)pState.TotalNumberOfItemsInDB / pState.ItemsPerPage.Value, 0, MidpointRounding.AwayFromZero);
                    return pState.With(z => z.NumberOfPages = new IntMore0Less65535Exclsv() { Value = res });
                };

        private static Func<PaginatorState, PaginatorState>
            IsValidLeft() => (pState) =>
                {
                    return pState.With(z => z.IsValidLeft = pState.CurrentPage.Value > 1);
                };

        private static Func<PaginatorState, PaginatorState>
            IsValidLeftMore() => (pState) =>
                 {
                     return pState.With(z => z.IsValidLeftMore = pState.CurrentPage.Value - pState.PagesToSkip.Value > 0);
                 };

        private static Func<PaginatorState, PaginatorState>
             IsValidRight() => (pState) =>
                 {
                     return pState.With(z => z.IsValidRight = pState.CurrentPage.Value < pState.NumberOfPages.Value);
                 };

        private static Func<PaginatorState, PaginatorState>
             IsValidRightMore() => (pState) =>
                 {
                     return pState.With(z => z.IsValidRightMore = pState.CurrentPage.Value + pState.PagesToSkip.Value < pState.NumberOfPages.Value);
                 };

        private static Func<PaginatorState, bool>
           IsValidItemsPerPage() => (pState) =>
                 {
                     return false;
                 };

        private static Func<PaginatorState, bool>
            IsValidPagesToSkip() => (pState) =>
                 {
                     return false;
                 };
    }
}

File: Paginator/PaginatorScope.fs
namespace Paginator

module PaginatorScope =
    open System
    open Paginator.Types
    
    let private GetNumberOfPages pState =
        let tmp = System.Math.Round( double pState.TotalNumberOfItemsInDB/double pState.ItemsPerPage.Value, MidpointRounding.AwayFromZero) |> uint16;
        {pState with NumberOfPages = IntMore0Less65535Exclsv tmp}

    let private IsValidLeft pState = 
        {pState with IsValidLeft = pState.CurrentPage.Value > 1us}

    let private IsValidLeftMore pState = 
        let tmp = if pState.PagesToSkip.Value > pState.CurrentPage.Value 
                    then false 
                  elif pState.CurrentPage.Value - pState.PagesToSkip.Value > 0us
                    then true
                  else false
       
        {pState with IsValidLeftMore = tmp}

    let private IsValidRight pState = 
        {pState with IsValidRight = pState.CurrentPage.Value < pState.NumberOfPages.Value}

    let private IsValidRightMore pState = 
        {pState with IsValidRightMore = pState.CurrentPage.Value + pState.PagesToSkip.Value < pState.NumberOfPages.Value}

    let private GetLeftIndex pState = 
        let tmp = ((pState.CurrentPage.Value - 1us) * pState.ItemsPerPage.Value) |> uint32
        {pState with LeftIndexInclsv  = tmp}

    let private GetRightIndex pState = 
        let tmp = (pState.CurrentPage.Value * pState.ItemsPerPage.Value  - 1us) |> uint32
        {pState with RightIndexInclsv  = tmp }

    let private  NextState pState =
         pState
         |> GetNumberOfPages  
         |> IsValidLeft  
         |> IsValidLeftMore  
         |> IsValidRight  
         |> IsValidRightMore
         |> GetLeftIndex
         |> GetRightIndex
   
    let Init pCurrentPage
             pItemsPerPageList
             pPagesToSkipList
             pTotalNumberOfItemsInDB =

         NextState   {NumberOfPages  = IntMore0Less65535Exclsv(1)
                      CurrentPage    = pCurrentPage
                      ItemsPerPage   = pItemsPerPageList
                      PagesToSkip    = pPagesToSkipList

                      TotalNumberOfItemsInDB = pTotalNumberOfItemsInDB
                      LeftIndexInclsv  = 0u
                      RightIndexInclsv  = uint32 (pItemsPerPageList.Value - 1us)

                      IsValidLeft      = false
                      IsValidLeftMore  = false
                      IsValidRight     = false
                      IsValidRightMore = false}

    let GoRight pState = 
        NextState {pState with CurrentPage = IntMore0Less65535Exclsv (pState.CurrentPage.Value + 1us)}

    let GoLeft pState = 
        NextState {pState with CurrentPage = IntMore0Less65535Exclsv (pState.CurrentPage.Value - 1us)}

    let GoRightMore pState = 
        NextState {pState with CurrentPage = IntMore0Less65535Exclsv (pState.CurrentPage.Value + pState.PagesToSkip.Value)}

    let GoLeftMore pState = 
        NextState {pState with CurrentPage = IntMore0Less65535Exclsv (pState.CurrentPage.Value - pState.PagesToSkip.Value)}   